### PR TITLE
feat: chat with compositeBrain ( with/out streaming)

### PR DESCRIPTION
**Pull request by gozineb**
# DONE

- generate_stream, generate and save answer in BE

# TODO

  -  Create an intermediary make_streaming_recursive_tool_calls async function 
  - Save intermediary answers in new message logs column then fetch and display in front

### CODE REVIEW

File: backend/modules/chat/controller/chat_routes.py [hunk 1]
```diff
@@ -158,6 +157,7 @@ async def create_question_handler(
     try:
         check_user_requests_limit(current_user)
         is_model_ok = (chat_question).model in user_settings.get("models", ["gpt-3.5-turbo"])  # type: ignore
+        print("HEELLOOOOO")
```
Comment (mamadoudicko) [inline backend/modules/chat/controller/chat_routes.py:160]: Hellowww

File: backend/llm/qa_headless.py [hunk 2]
```diff
@@ -102,7 +101,7 @@ def _create_prompt_template(self):
         return CHAT_PROMPT
 
     def generate_answer(
-        self, chat_id: UUID, question: ChatQuestion
+        self, chat_id: UUID, question: ChatQuestion, save_answer: bool = True
```
Comment (mamadoudicko) [inline backend/llm/qa_headless.py:104]: Do you think we should update `QAInterface` too ?

File: backend/llm/composite_brain_qa.py [hunk 3]
```diff
@@ -0,0 +1,611 @@
...
+                    "brain": {
+                        "type": "string",
+                        "description": "Brain details, need to check type too",
+                    },
+                    "question": {
+                        "type": "string",
+                        "description": "Question to ask the brain",
+                    },
+                },
+                "required": ["brain_id", "question"],
```
Comment (mamadoudicko) [inline backend/llm/composite_brain_qa.py:42]: Not `brain` instead of `brain_id` here ?

File: backend/llm/composite_brain_qa.py [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+                    CreateChatHistory(
+                        **{
+                            "chat_id": chat_id,
+                            "user_message": question.question,
+                            "assistant": response.assistant,
+                            "brain_id": question.brain_id,
+                            "prompt_id": self.prompt_to_use_id,
+                        }
+                    )
+                )
```
Comment (mamadoudicko) [inline backend/llm/composite_brain_qa.py:149]: Maybe directly calling generate_answer of `HeadlessQA` with `save_answer = true` ?

Comment (mamadoudicko) [reply backend/llm/composite_brain_qa.py:149]: Something like :
```python

    def generate_answer(
        self, chat_id: UUID, question: ChatQuestion, save_answer: bool
    ) -> str:
        connected_brains = brain_service.get_connected_brains(self.brain_id)
        if not connected_brains:
            return HeadlessQA(
                chat_id=chat_id,
                model=self.model,
                max_tokens=self.max_tokens,
                temperature=self.temperature,
                streaming=self.streaming,
                prompt_id=self.prompt_id,
            ).generate_answer(chat_id, question, save_answer=save_answer)
```

File: backend/llm/composite_brain_qa.py [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+                }
+            )
+
+        tools = []
+        available_functions = {}
+
+        connected_brains_details = {}
+        for brain_id in connected_brains:
+            brain = brain_service.get_brain_by_id(brain_id)
+            if brain == None:
```
[the remark this answers is not in the document]
Comment (mamadoudicko) [reply backend/llm/composite_brain_qa.py:183]: https://stackoverflow.com/a/26611

File: backend/llm/composite_brain_qa.py (hunk 3 above)
Comment (gozineb) [reply backend/llm/composite_brain_qa.py:42]: yes

**Review by mamadoudicko — APPROVED** [review]

**Pull request merged**